Bar memory update develop (#4168)
RadeonOpenCompute/ROCm: docs/how-to/Bar-Memory.rst

.. meta::
   :description: Learn about BAR configuration in AMD GPUs and ways to troubleshoot physical addressing limit
   :keywords: BAR memory, MMIO, GPU memory, Physical Addressing Limit, AMD, ROCm

**************************************
Troubleshoot BAR access limitation
**************************************
Direct Memory Access (DMA) to PCIe devices using Base Address Registers (BARs) can be restricted due to physical addressing limits. These restrictions can result in data access failures between the system components. Peer-to-peer (P2P) DMA is used to access resources such as registers and memory between devices. PCIe devices need memory-mapped input/output (MMIO) space for DMA, and these MMIO spaces are defined in the PCIe BARs.

> These BARs are a set of 32-bit or 64-bit registers that are used to define the memory space that PCIe devices can access. The CPU and other system devices also use these to access the resources of the PCIe devices. P2P DMA only works when one device can directly access the local BAR memory of another. If the BAR memory is more than the physical addressing limit of the device, the device will not be able to access the remote BAR. 

Reviewer: "the memory space that PCIe devices can access"
This is not correct.  It should say:
"the resources that PCIe devices provide"

> These BARs are a set of 32-bit or 64-bit registers that are used to define the memory space that PCIe devices can access. The CPU and other system devices also use these to access the resources of the PCIe devices. P2P DMA only works when one device can directly access the local BAR memory of another. If the BAR memory is more than the physical addressing limit of the device, the device will not be able to access the remote BAR. 

Reviewer: " If the BAR memory is more than the physical addressing limit of the device, the device will not be able to access the remote BAR. "
I would rephrase this like so:
"If the location of a BAR memory is above the physical addressing limit of a device, the device will not be able to access that BAR.   This could be a device's own BAR or the BAR of another device on the system."

To handle any BAR access issues that might occur, you need to be aware of the physical address limitations of the devices and understand the :ref:`BAR configuration of AMD GPUs <bar-configuration>`. This information is important when you need to set up additional MMIO apertures for PCIe devices in the system's physical address space.

Handling physical address limitation
=============================================

> When a system boots, the system BIOS allocates the physical address space for the components in the system, including system memory and MMIO apertures. There are two MMIO apertures: one allocated less than 4 GB of physical address space for 32-bit compatibility, and one more than 4 GB for devices needing more space.

Reviewer: "There are two MMIO apertures: one allocated less than 4 GB of physical address space for 32-bit compatibility, and one more than 4 GB for devices needing more space."
I'd rephrase like so:
"On modern 64-bit platforms, there are generally two or more MMIO apertures: one located below 4 GB of physical address space for 32-bit compatibility, and one or more above 4 GB for devices needing more space."

There are two ways to handle this:

* Ensure that the high MMIO aperture is within the physical addressing limits of the devices in the system. For example, if the devices have a 44-bit physical addressing limit, set the ``MMIO High Base`` and ``MMIO High size`` options in the BIOS such that the aperture is within the 44-bit address range, and ensure that the ``Above 4G Decoding`` option is Enabled.  

* Enable the Input-Output Memory Management Unit (IOMMU). When the IOMMU is enabled in non-passthrough mode, it will create a virtual I/O address space for each device on the system. It also ensures that all virtual addresses created in that space are within the physical addressing limits of the device. For more information on IOMMU, see :doc:`../conceptual/iommu`. 

.. _bar-configuration:

BAR configuration for AMD GPUs
================================================

> The following table shows how the BARs are configured for AMD GPUs that have a 44-bit physical address and 48-bit GPU virtual address limit. 

Reviewer: "that have a 44-bit physical address and 48-bit GPU virtual address limit"
I would drop this part.  The BAR configuration is the same regardless of what the GPU's physical addressing limits are.  Also I wouldn't mention the GPU virtual addressing limit here, it will just confuse things as it's not related to this.

.. meta::
   :description: Learn about BAR configuration in AMD GPUs and ways to troubleshoot physical addressing limit
   :keywords: BAR memory, MMIO, GPU memory, Physical Addressing Limit, AMD, ROCm

**************************************
Troubleshoot BAR access limitation
**************************************
Direct Memory Access (DMA) to PCIe devices using Base Address Registers (BARs) can be restricted due to physical addressing limits. These restrictions can result in data access failures between the system components. Peer-to-peer (P2P) DMA is used to access resources such as registers and memory between devices. PCIe devices need memory-mapped input/output (MMIO) space for DMA, and these MMIO spaces are defined in the PCIe BARs.

> These BARs are a set of 32-bit or 64-bit registers that are used to define the resources that PCIe devices provide. The CPU and other system devices also use these to access the resources of the PCIe devices. P2P DMA only works when one device can directly access the local BAR memory of another. If the memory address of a BAR memory is more than the physical addressing limit of a device, the device will not be able to access that BAR. This could be the device's own BAR or the BAR of another device in the system.

Reviewer 2: ```suggestion
These BARs are a set of 32-bit or 64-bit registers that are used to define the resources that PCIe devices provide. The CPU and other system devices also use these to access the resources of the PCIe devices. P2P DMA only works when one device can directly access the local BAR memory of another. If the memory address of a BAR memory exceeds the physical addressing limit of a device, the device will not be able to access that BAR. This could be the device's own BAR or the BAR of another device in the system.
```

> If the BAR memory is more than the physical addressing limit of the device, the device will not be able to access the remote BAR. 

Reviewer 2: ```suggestion
If the BAR memory exceeds the physical addressing limit of the device, the device will not be able to access the remote BAR. 
```

Reviewer 2: i think i previously suggested more than instead of over. Not sure that worked

> To handle any BAR access issues that might occur, you need to be aware of the physical address limitations of the devices and understand the :ref:`BAR configuration of AMD GPUs <bar-configuration>`. This information is important when you need to set up additional MMIO apertures for PCIe devices in the system's physical address space.

Reviewer 2: ```suggestion
To handle any BAR access issues that might occur, you need to be aware of the physical address limitations of the devices and understand the :ref:`BAR configuration of AMD GPUs <bar-configuration>`. This information is important when setting up additional MMIO apertures for PCIe devices in the system's physical address space.
```
